src/item.cpp

```diff
@@ -9315,24 +9315,27 @@ ret_val<bool> item::is_compatible( const item &it ) const
     return contents.is_compatible( it );
 }
 
-ret_val<bool> item::can_contain( const item &it ) const
+ret_val<bool> item::can_contain( const item &it, const bool nested ) const
 {
     if( this == &it ) {
         // does the set of all sets contain itself?
         return ret_val<bool>::make_failure();
     }
+    if( nested && !this->is_container() ) {
+        return ret_val<bool>::make_failure();
+    }
     // disallow putting portable holes into bags of holding
     if( contents.bigger_on_the_inside( volume() ) &&
         it.contents.bigger_on_the_inside( it.volume() ) ) {
         return ret_val<bool>::make_failure();
     }
     for( const item *internal_it : contents.all_items_top( item_pocket::pocket_type::CONTAINER ) ) {
-        if( internal_it->contents.can_contain_rigid( it ).success() ) {
+        if( internal_it->can_contain( it, true ).success() ) {
```

Comment:
Does this work properly? It seems to me this could give false positives when the item fits in a nested pocket but not the parent item.

Reply:
I _think_ this makes sense.
I made this change thinking about pocket restrictions with something like the following scenario (using abstract values for the purpose of explanation):

Backpacks have a special pocket for water bottles with a min-length of 2".
If pocket contains a water bottle without a min-length restriction, you could try to fit some nails in there instead of the backpack pocket.

If you mean `can_contain` should only function for one item (and not nested items), I can kind of agree and it'd have to be a separate method (similar to `item_pocket::best_pocket_in_contents`).

Reply:
That's a valid case, but I'm worried about removing the rigid part. For example if you have a plastic bag inside a jeans pocket, wouldn't that give a success for anything that fits in a plastic bag, even if it doesn't fit in a jeans pocket?

Reply:
That should be handled because of the `nested` boolean with this recursive calling of `can_contain`.
If you look at the bottom of this method, it checks if `nested` and determines if it should check rigid or not in that case.

Reply:
To put it better, the issue here is that there are some pocket properties that can be bypassed by nested pockets and some that can't. An internal item can be watertight if the parent pocket isn't, but a soft internal item can't contain a larger volume than the parent pocket. There needs to be separate handling for both types of properties.

Reply:
> That should be handled because of the `nested` boolean with this recursive calling of `can_contain`. If you look at the bottom of this method, it checks if `nested` and determines if it should check rigid or not in that case.

Oh yeah, you're right, I was reading that wrong.

Reply:
Could've been me as well! I'm not always the best at explaining 👍

Reply:
You can mark this as resolved btw.